src/replication.c

```diff
@@ -1270,6 +1288,56 @@ void replconfCommand(client *c) {
                 }
             }
             sdsfreesplitres(filters, filter_count);
+        } else if (!strcasecmp(c->argv[j]->ptr, "end-offset")) {
+            /* REPLCONF end-offset is used to identify the client wants
+            *  the snapshot end offset once snapshot is sent. */
+            long start_with_offset = 0;
+            if (getRangeLongFromObjectOrReply(c, c->argv[j +1],
+                    0, 1, &start_with_offset,NULL) != C_OK) {
+                return;
+            } 
+            if (start_with_offset == 1) {
+                c->flags |= CLIENT_RDB_CHANNEL_RDBCHAN;
+            } else {
+                c->flags &= ~CLIENT_RDB_CHANNEL_RDBCHAN;
+            }
+        } else if (!strcasecmp(c->argv[j]->ptr, "psync-only")) {
+            /* REPLCONF psync-only is used to identify the client wants
+             * to partial sync if possible. If not, do not auto 
+             * trigger full sync. We use it during rdb channel sync
+             * to mark the main connection. */
+            long psync_only = 0;
+            if (getRangeLongFromObjectOrReply(c, c->argv[j +1],
+                    0, 1, &psync_only, NULL) != C_OK) {
+                return;
+            } 
+            if (psync_only == 1) {
+                if (!server.repl_diskless_sync) {
+                    /* When the primary uses disk for full sync, replicas can usually join during the time the 
+                     * primary saves the database to disk. RDB-channel-sync, however, does not allow replicas 
+                     * to join the primary since the COB does not keep the needed replication data. To avoid 
+                     * making the primary create a new snapshot for each replica, we forbid rdb-channel-sync
+                     * along with primary disk-based sync */
+                    addReplyErrorFormat(c,"Rdb channel sync is not allowed when repl-diskless-sync disabled on primary");
+                    return;
+                }
+                c->flags |= CLIENT_RDB_CHANNEL_PSYNCCHAN;
+                c->slave_capa |= SLAVE_CAPA_PSYNC2;
+            } else {
+                c->flags &= ~CLIENT_RDB_CHANNEL_PSYNCCHAN;
+            }
+        } else if (!strcasecmp(c->argv[j]->ptr, "connected")) {
```

Comment:
Do we need this step? My understanding is that the replica will start sending "REPLCONF ACK" once it is only, why do we need a new mechanism to indicate we are online?

Reply:
Yes, we can use replconf ack instead of replconf connect. I missed that.